### DISCUSSION TITLE
Is it possible to make shift layer and key position layer orthogonal?

### DISCUSSION THREAD

**Discussion by glyh** [kmonad/kmonad#528] [Q&A]
As title.

For example, on top of the default QWERTY layer, I have COLEMAK-dk, and some other keyboard layout. They only affect, say, alphas. 

I also want to remap some specific keys, specifically, I want `shift + [` to be `(`, and `shift + ]` to be `)`. But I can't really fall through other keys, because o.w. they would not be capitalized. 

So is it possible to make them orthogonal? Or do I have to repeat every key layout twice? Thanks a lot!

If this is not possible yet, I guess it would be good if we have some kind of functions that manipulate the input when falling through, working like this I guess?

```lisp
(defsrc 
   a)
(deflayer shift 
    #f)
(deflayer colemak-dh
    x)
(defn f [key] (captialize key)) 
;; Functions are like "generalized transparent(_)", while `_` can be represented as the identity function. 
;; Literal keys can be represented as constant functions.
```
Though I know there's significant work needed for this. I guess this can be done as a separate project implemented with true lisp?

**Comment by MuhammedZakir** [comment]
I am hoping things like this to be possible after keycode refactoring is done. I personally don't want a full lisp. The reason is explained in the below link, and IMHO, it would only increase unneeded complexity for both devs and users.

See https://github.com/kmonad/kmonad/issues/237 for more details.

Edit: When I said "possible after keycode refactoring", what I meant was that it will be (hopefully) easier to implement this, and not that this will be possible just by refactoring keycode stuff.
